File: physm-js/src/Spring.ts
export interface SpringJson {
  stiffness: number;
}

/**
 * A spring acting on the frame's own coordinate, slack at zero.
 *
 * The restoring generalised force is `-stiffness * q`: a torque on a
 * rotational joint, and a force along the axis of a track. It is conservative
 * -- `U = stiffness * q^2 / 2` -- so it enters the potential with gravity
 * rather than the dissipative terms, which is what `docs/algorithm.md` means
 * by "the frame's own coordinate is the whole of its input".
 *
 * **A node rather than a prop on the frame**, for the reasons a frame's
 * `Weight`s are: there may be several, they are separate things a person adds
 * and removes, and they carry their own parameters. While every spring is
 * linear the multiplicity buys nothing -- two of them sum exactly to one --
 * and it buys everything the moment one is not, which is where a stop that
 * engages past a threshold, or a stiffness that rises with the angle, would
 * go. `docs/issues/0016/12-wishlist.md` asks for exactly that.
 *
 * It is also what keeps a *joint* spring and a spring anchored to something
 * else distinguishable. They are different devices -- one belongs to a
 * frame's coordinate and one names two ends -- and a pair of numbers on the
 * frame is a shape in which that cannot be said.
 */
export default class Spring {
  readonly stiffness: number;

  constructor(stiffness = 0) {
    this.stiffness = stiffness;
  }

  /**
   * What this spring contributes to its frame's generalised force.
   *
   * Asked of the spring rather than computed by the solver, so that what a
   * spring is slack *toward* stays the spring's own business: the solver hands
   * over the coordinate and adds up what it gets back.
   */
  force(q: number): number {
    return -this.stiffness * q;
  }

  toJsonObj(): SpringJson {
    return { stiffness: this.stiffness };
  }
}

File: physm-rs/src/spring.rs
use crate::json;
use crate::Error;

/// A spring acting on its frame's own coordinate, slack at zero.
///
/// The restoring generalised force is `-stiffness * q`: a torque on a
/// rotational joint, a force along the axis of a track. Conservative -- it
/// enters `U` with gravity rather than the dissipative terms.
///
/// A list on the frame rather than a number on it, mirroring `physm-js`'s
/// `Spring`: there may be several, and while each is linear their sum is one
/// spring of the summed stiffness. That stops being true the moment one is
/// not, which is where a stop that engages past a threshold would go.
#[derive(Debug, PartialEq)]
pub struct Spring {
    pub stiffness: f64,
}

impl Spring {
    pub fn new(stiffness: f64) -> Self {
        Self { stiffness }
    }

    /// What this spring contributes to its frame's generalised force.
    pub fn get_force(&self, q: f64) -> f64 {
        -self.stiffness * q
    }

    pub fn from_json_value(value: &serde_json::Value) -> Result<Self, Error> {
        let obj = json::value_to_json_obj(value)?;
        Ok(Spring {
            stiffness: json::map_obj_item_or_default(obj, "stiffness", json::value_to_f64)?,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_force() {
        assert_abs_diff_eq!(Spring::new(4.).get_force(0.5), -2.);
        assert_abs_diff_eq!(Spring::new(4.).get_force(-0.5), 2.);
        assert_abs_diff_eq!(Spring::new(0.).get_force(0.5), 0.);
    }

    #[test]
    fn test_from_json_value() {
        let parse = |json: &str| {
            Spring::from_json_value(&serde_json::from_str::<serde_json::Value>(json).unwrap())
                .unwrap()
        };

        assert_eq!(parse(r#"{"stiffness": 6.25}"#), Spring::new(6.25));

        // A spring that states nothing is slack, rather than failing to parse.
        assert_eq!(parse(r#"{}"#), Spring::new(0.));
    }
}
